### PR TITLE
Some DAA Fixes As They Come Up

### PR DESCRIPTION
**Pull request by flagrama**

### CODE REVIEW

**Comment by flagrama** [comment]
[TOXAPEX_54](https://trello.com/c/tLeyVDcN/488-toxapex54)

**Comment by axpendix** [comment]
Hey, what is the status of this one?

**Comment by flagrama** [comment]
A couple more cards had issues or things I need to check. Shouldn't actually interfere with #883 and if it does, I can rebase.

File: src/tcgwars/logic/impl/gen8/DarknessAblaze.groovy [hunk 1]
```diff
@@ -1,4 +1,5 @@
-package tcgwars.logic.impl.gen8;
+package tcgwars.logic.impl.gen8
```
Comment (flagrama) [inline src/tcgwars/logic/impl/gen8/DarknessAblaze.groovy:1]: I guess IntelliJ didn't like all these imports. I can undo this if needed.

File: src/tcgwars/logic/impl/gen8/DarknessAblaze.groovy [hunk 2]
```diff
@@ -14,21 +15,11 @@ import static tcgwars.logic.effect.EffectPriority.*
...
-import tcgwars.logic.effect.ability.*;
-import tcgwars.logic.effect.advanced.*;
-import tcgwars.logic.effect.basic.*;
-import tcgwars.logic.effect.blocking.*;
-import tcgwars.logic.effect.event.*;
-import tcgwars.logic.effect.getter.*;
+import tcgwars.logic.effect.ability.*
+import tcgwars.logic.effect.basic.*
 import tcgwars.logic.effect.special.*;
 import tcgwars.logic.util.*;
```
Comment (flagrama) [inline src/tcgwars/logic/impl/gen8/DarknessAblaze.groovy:25]: I guess IntelliJ didn't like all these imports. I can undo this if needed.

File: src/tcgwars/logic/impl/gen8/DarknessAblaze.groovy [hunk 3]
```diff
@@ -1390,9 +1387,9 @@ public enum DarknessAblaze implements LogicCardInfo {
         bwAbility "Primal Law", {
           text "If this Pokémon is your Active Pokémon, your opponent can’t play any Pokémon from their hand to evolve their Pokémon."
           delayedA {
-            before PLAY_EVOLUTION, {
-              if (self.active && bg.currentTurn == self.owner.opposite && ef.cardToPlay.player.pbg.hand.contains(ef.cardToPlay)) {
-                wcu "$self.name's Primal Law prevents playing Pokémon from your hand to evolve your Pokémon"
+            before EVOLVE_STANDARD, {
+              if (self.active && bg.currentTurn == self.owner.opposite && (ef as EvolveStandard).evolutionCard.player.pbg.hand.contains(ef.evolutionCard)) {
+                wcu "$self's Primal Law prevents playing Pokémon from your hand to evolve your Pokémon"
```
Comment (starg09) [inline src/tcgwars/logic/impl/gen8/DarknessAblaze.groovy:1392]: Really minor detail, but it seems two warning conflict with each other when directly playing an evolution from hand.

Old Rare Candy seems to skip this prevention still, but that's not really important (and shouldn't even be a problem for Unlimited, whenever common code is implemented for it like with Pokemon Catcher)

Ditto Prism Star seems to bypass this effect. Same with Ascension DNA in Eevee-GX, and Quick-Ripening Herb in Meganium (Lost Thunder).

Comment (starg09) [reply src/tcgwars/logic/impl/gen8/DarknessAblaze.groovy:1392]: > Really minor detail, but it seems two warning conflict with each other when directly playing an evolution from hand

Oh, forgot to pass a screenshot I took of that:

image

Comment (flagrama) [reply src/tcgwars/logic/impl/gen8/DarknessAblaze.groovy:1392]: The stacking issue seems to be something Vaadin 14 fixes https://github.com/vaadin/framework/issues/676#issuecomment-544166861

Playing an evolution directly from hand shouldn't cause stacked notifications now, however Rare Candy does. A lot. 4 of them. 2 of which are empty currently due to a bug in the main repo.

File: src/tcgwars/logic/impl/gen8/DarknessAblaze.groovy [hunk 4]
```diff
@@ -1390,12 +1387,22 @@ public enum DarknessAblaze implements LogicCardInfo {
         bwAbility "Primal Law", {
           text "If this Pokémon is your Active Pokémon, your opponent can’t play any Pokémon from their hand to evolve their Pokémon."
           delayedA {
-            before PLAY_EVOLUTION, {
-              if (self.active && bg.currentTurn == self.owner.opposite && ef.cardToPlay.player.pbg.hand.contains(ef.cardToPlay)) {
-                wcu "$self.name's Primal Law prevents playing Pokémon from your hand to evolve your Pokémon"
+            def warnAndPrevent = {
+              if (self.active && bg.currentTurn == self.owner.opposite) {
+                wcu "$self's Primal Law prevents playing Pokémon from your hand to evolve your Pokémon"
                 prevent()
```
Comment (axpendix) [inline src/tcgwars/logic/impl/gen8/DarknessAblaze.groovy:1393]: Not sure this works here 🤔 `prevent()` is a function of the `before` closure's `delegate`. Have you tested this?

File: src/tcgwars/logic/impl/gen8/DarknessAblaze.groovy (hunk 2 above)
Comment (axpendix) [reply src/tcgwars/logic/impl/gen8/DarknessAblaze.groovy:25]: I think it's better to revert these

File: src/tcgwars/logic/impl/gen8/DarknessAblaze.groovy (hunk 4 above)
Comment (starg09) [reply src/tcgwars/logic/impl/gen8/DarknessAblaze.groovy:1393]: Tested it yesterday and it seemed to work. Just tested with the last changes and debugging, prevent executes without issue 👍

File: src/tcgwars/logic/impl/gen8/DarknessAblaze.groovy (hunk 3 above)
Comment (starg09) [reply src/tcgwars/logic/impl/gen8/DarknessAblaze.groovy:1392]: > however Rare Candy does
Doesn't seem to pop em up here now, lol.

Hey, if it works it works 😛

File: src/tcgwars/logic/impl/gen8/DarknessAblaze.groovy (hunk 4 above)
Comment (flagrama) [reply src/tcgwars/logic/impl/gen8/DarknessAblaze.groovy:1393]: I have, and it seems to work fine. Pretty sure this closure's delegate is the before, then the before's delegate has a prevent, so it kind of just bubbles up the tree.

**Pull request merged**